### PR TITLE
docs: Noise comparison

### PR DESCRIPTION
**Pull request by GKD-stack**
This walks the user through the different types of noise, explores use cases where one may be better suited, and then demonstrates how to specify them in the context API. 

This should be merged first: https://github.com/opendp/opendp/pull/1812
https://github.com/opendp/opendp/pull/1834
https://github.com/opendp/opendp/pull/1836

this is a rough draft and I did have these questions: 
1. is rho generally 0.5?

these are polars questions mostly
1. why does adding a column with with_columns and then casting it to a string cause a dtype error with defining the margin? 
2. trying to do pl.sum also causes a margin error
`(context.query().fill_nan(40.0).fill_null(40.0).group_by("ILOSTAT").agg(pl.sum("HWUSUAL").dp.laplace().alias("Laplace Noise"))).release().collect()`

`MakeMeasurement("Failed to find margin for {"ILOSTAT"}")`

but the margin is defined here

`context = dp.Context.compositor(
    data=df,
    privacy_unit=dp.unit_of(contributions=36),
    privacy_loss=dp.loss_of(epsilon=1.0),
    split_evenly_over=10,
    margins={
        ("ILOSTAT", ): dp.Margin(public_info="keys", max_partition_length=60_000_000),
        ("HWUSUAL", ): dp.Margin(public_info="keys", max_partition_length=60_000_000),
        (): dp.Margin(max_partition_length=60_000_000),
    },
)`

(The HWUSUAL margin may not be included in the version uploaded, but it doesn't work either way)

Thanks!

### CODE REVIEW

File: docs/Noise.ipynb
Comment (mccalluc) [inline docs/Noise.ipynb:330]: Repeating a comment made elsewhere, we want notebooks to run without warnings... and if that's not possible, we should comment to explain why, and have an issue.

File: docs/Noise.ipynb
Comment (mccalluc) [inline docs/Noise.ipynb:374]: Great! But where does this come from? Source URL in comment.

**Review by mccalluc — COMMENTED** [review]
I think Mike also would have comments about the presentation of this information. I would defer to his judgement, but my comments might have some useful points on writing documentation more generally, even if you don't end up using them here.

**Comment by mccalluc** [comment]
Also, note
- #1774

We'll want this in a new directory in the docs. I'll make a separate PR for that, just to get the ball rolling.

File: docs/Noise.ipynb
Comment (Shoeboxam) [reply docs/Noise.ipynb:23]: More specifically, it chooses based on the privacy measure. If pure-DP, then laplace, if zCDP, then gaussian.

File: docs/Noise.ipynb
Comment (Shoeboxam) [reply docs/Noise.ipynb:40]: There is actually a notion of "strength" of privacy measures. I would put this more along the lines that pure DP is a stronger definition of privacy than approximate-DP, and since the gaussian mechanism cannot satisfy pure-DP, it's categorically different which can make it difficult to compare the two mechanisms fairly.

Not so relevant to the notebook, but more broadly, you can think of the strength of a privacy measure as a kind of partial ordering. All pure-DP guarantees can be converted into approx-DP guarantees, but not vice-versa. This is why pure-DP is "stronger" than approx-DP. [The following graph is not complete](https://mermaid.live/edit#pako:eNp1kLsKgzAUhl8lHOhWKWR0KEg7dgjpZuIQ9NgKaiQk9CK-e62JthaaKf8lfD_pIdcFQgxlrW_5VRlLTly2ZDxMMGcwI1G0J6l4Ho4s80E6WVxwbB_Vyku84F4IZnRZ1bgro-XtZkPYXA0YKpKuM_oeSLOMvonUI5dshaYr9k_n_xRGP0OCMxmMLlPTMGkxeNgBW2jQNKoqxq_r37EEe8UGJcTjtcBSudpKkO0wVpWz-vxoc4itcbgF1xXK4rFSF6Mabw4vpWVwFw), but gives a sense of the relationships between some of the common privacy measures.

File: docs/Noise.ipynb
Comment (Shoeboxam) [reply docs/Noise.ipynb:47]: I agree, we definitely shouldn't be pulling assets from other resources.

```python
import matplotlib.pyplot as plt
import numpy as np

def laplace_pdf(b, mu=0):
    return lambda x: np.exp(-abs(x - mu) / b) / (2 * b)

def gaussian_pdf(b, mu=0):
    return lambda x: np.exp(-((x - mu) / b) ** 2 / 2) / (np.sqrt(2 * np.pi) * b)


scale = 1.
cont_support = np.linspace(-5, 5, num=1000)
plt.plot(cont_support, laplace_pdf(scale)(cont_support), label="M(\cdot) ~ Laplace")
plt.plot(cont_support, gaussian_pdf(scale)(cont_support), label="M(\cdot) ~ Gaussian")
plt.xlabel("support ($y$)")
plt.ylabel("$\Pr[M(0) = y]$")
plt.title("Laplace vs. Gaussian Distributions")
plt.legend()
plt.show()
```

File: docs/Noise.ipynb
Comment (Shoeboxam) [reply docs/Noise.ipynb:53]: Agree with Chuck about using newlines, it makes it much easier to review.

The gaussian mechanism doesn't necessarily perform better than the laplace mechanism for a vector-- in the case where the L1 sensitivity is equal to the L2 sensitivity, the laplace mechanism provides overall lower variance.

Gaussian utility is better particularly in the cases where:
1. the L2 sensitivity is smaller than L1 (and discussing when this might be the case would be interesting)
2. there are many compositions

File: docs/Noise.ipynb
Comment (GKD-stack) [reply docs/Noise.ipynb:124]: im leaving the adding a link to the introduction page HTML comment in markdown as with the other notebooks

File: docs/Noise.ipynb
Comment (GKD-stack) [reply docs/Noise.ipynb:258]: ok, im leaving the todo in the markdown to insert a link

File: docs/Noise.ipynb
Comment (GKD-stack) [reply docs/Noise.ipynb:305]: nope, took it out for hours usual

**Review by mccalluc — APPROVED** [review]
Tests pass, and the explanation seems good. There might be more edits before publication, but I'll go ahead with merge.

File: docs/source/getting-started/assessing-utility/noise.ipynb
Comment (vikrantsinghal) [inline docs/source/getting-started/assessing-utility/noise.ipynb:21]: "...different types of distributions available to sample the noise from, including the Laplacian and the Gaussian distributions."

File: docs/source/getting-started/assessing-utility/noise.ipynb
Comment (vikrantsinghal) [inline docs/source/getting-started/assessing-utility/noise.ipynb:27]: Mention just before this code snippet what this is doing. Could be as simple as, "As always we do the blah blah thing in the following."

File: docs/source/getting-started/assessing-utility/noise.ipynb
Comment (vikrantsinghal) [inline docs/source/getting-started/assessing-utility/noise.ipynb:53]: They don't really appear similar, to be honest. They have pretty different tail properties, so can't really say that.

File: docs/source/getting-started/assessing-utility/noise.ipynb
Comment (vikrantsinghal) [inline docs/source/getting-started/assessing-utility/noise.ipynb:56]: - "satisfy (pure) ε-DP, whereas..."
- What does precise guarantee mean? I think approx DP is also a "precise" guarantee. Might just want to say that Laplace provides a "stronger" privacy guarantee.

File: docs/source/getting-started/assessing-utility/noise.ipynb
Comment (vikrantsinghal) [inline docs/source/getting-started/assessing-utility/noise.ipynb:105]: "when:"

File: docs/source/getting-started/assessing-utility/noise.ipynb
Comment (vikrantsinghal) [inline docs/source/getting-started/assessing-utility/noise.ipynb:158]: Again, this looks untidy. Enclose it in a box or something because this is either code or some output.

File: docs/source/getting-started/assessing-utility/noise.ipynb
Comment (vikrantsinghal) [inline docs/source/getting-started/assessing-utility/noise.ipynb:322]: One issue here is that we're talking about ρ-zCDP in this line without giving any context on that. Previously, we only talk about approx DP, but here in the tutorial, we don't really do that, but use zCDP, instead.

Comment (GKD-stack) [reply docs/source/getting-started/assessing-utility/noise.ipynb:322]: I took out this line from the introduction: Otherwise, the Gaussian distribution may be used to satisfy zero concentrated DP...

is that what you meant?

**Pull request merged**
